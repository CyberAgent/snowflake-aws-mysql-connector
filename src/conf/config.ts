import * as fs from "fs";
import * as path from "path";

export interface Config {
  stackName: string;
  awsAccountId: string;
  awsRegion: string;
  lambdaFunctionName: string;
  lambdaEnvironmentVariables?: { [key: string]: string };
  vpcId: string;
  subnets: { id: string; availabilityZone: string }[];
  securityGroupIds: string[];
  snowflakeApiAwsIamUser: { arn: string; externalId: string }[];
  mysqlHost: string;
  mysqlUser: string;
  mysqlPassword: string;
  mysqlDatabase: string;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function isConfig(config: any): config is Config {
  return config.stackName !== undefined;
}

export function getConfig(target: string, _basePath?: string): Config {
  let basePath;
  if (_basePath) {
    basePath = _basePath;
  } else {
    basePath = "../conf-file";
  }
  const filePath = path.resolve(__dirname, `${basePath}/${target}.json`);
  const json = JSON.parse(fs.readFileSync(filePath, "utf8"));
  if (isConfig(json)) {
    if (json.lambdaEnvironmentVariables) {
      json.lambdaEnvironmentVariables["STACK_ENV"] = target;
    } else {
      json.lambdaEnvironmentVariables = { STACK_ENV: target };
    }
    return json;
  } else {
    throw new Error("config type check failed");
  }
}
