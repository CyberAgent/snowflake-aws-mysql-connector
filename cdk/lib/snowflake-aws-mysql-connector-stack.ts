import * as apigateway from "aws-cdk-lib/aws-apigateway";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as iam from "aws-cdk-lib/aws-iam";
import * as logs from "aws-cdk-lib/aws-logs";
import * as cdk from "aws-cdk-lib/core";
import { Config } from "../../src/conf/config";
import { Construct } from "constructs";

export interface SnowflakeAWSMySQLConnectorStackProps extends cdk.StackProps {
  readonly config: Config;
  readonly lambdaAssetPath: string;
  readonly lambdaLayerAssetPath: string;
}

export class SnowflakeAWSMySQLConnectorStack extends cdk.Stack {
  constructor(
    scope: Construct,
    id: string,
    props: SnowflakeAWSMySQLConnectorStackProps
  ) {
    super(scope, id, props);

    const lambdaVpc = ec2.Vpc.fromLookup(
      this,
      `${props.config.lambdaFunctionName}-vpc`,
      { vpcId: props.config.vpcId }
    );

    const lambdaSubnets: ec2.ISubnet[] = [];
    for (const s of props.config.subnets) {
      const subnet = ec2.Subnet.fromSubnetAttributes(
        this,
        `${props.config.lambdaFunctionName}-subnet`,
        {
          subnetId: s.id,
          availabilityZone: s.availabilityZone,
        }
      );
      lambdaSubnets.push(subnet);
    }

    const lambdaSecurityGroups: ec2.ISecurityGroup[] = [];
    for (const i in props.config.securityGroupIds) {
      const securityGroup = ec2.SecurityGroup.fromSecurityGroupId(
        this,
        `${props.config.lambdaFunctionName}-security-group`,
        props.config.securityGroupIds[i]
      );
      lambdaSecurityGroups.push(securityGroup);
    }

    const lambdaLayer = new lambda.LayerVersion(
      this,
      `${props.config.lambdaFunctionName}-layer`,
      {
        code: lambda.AssetCode.fromAsset(props.lambdaLayerAssetPath),
        compatibleRuntimes: [lambda.Runtime.NODEJS_16_X],
      }
    );
    const lambdaFunction = new lambda.Function(
      this,
      props.config.lambdaFunctionName,
      {
        functionName: props.config.lambdaFunctionName,
        code: lambda.Code.fromAsset(props.lambdaAssetPath),
        layers: [lambdaLayer],
        handler: "index.handler",
        runtime: lambda.Runtime.NODEJS_16_X,
        timeout: cdk.Duration.seconds(10),
        logRetention: logs.RetentionDays.THREE_DAYS,
        environment: props.config.lambdaEnvironmentVariables,
        vpc: lambdaVpc,
        vpcSubnets: lambdaVpc.selectSubnets({ subnets: lambdaSubnets }),
        securityGroups: lambdaSecurityGroups,
        allowPublicSubnet: true,
      }
    );

    const snowflakePrincipals: iam.PrincipalBase[] = [];
    const snowflakeExternalIds: string[] = [];
    const externalIds: Set<string> = new Set();
    const arns: Set<string> = new Set();
    for (const user of props.config.snowflakeApiAwsIamUser) {
      arns.add(user.arn);
      externalIds.add(user.externalId);
    }
    for (const arn of arns) {
      const principal = new iam.ArnPrincipal(arn);
      snowflakePrincipals.push(principal);
    }
    snowflakeExternalIds.push(...Array.from(externalIds));

    if (snowflakePrincipals.length == 0) {
      snowflakePrincipals.push(new iam.ServicePrincipal("ec2.amazonaws.com"));
    }
    if (snowflakeExternalIds.length == 0) {
      snowflakeExternalIds.push(
        "please fill in snowflakeApiAwsIamUser. this is filler value to avoid errors."
      );
    }

    const snowflakeRole = new iam.Role(
      this,
      `${props.config.lambdaFunctionName}-snowflake-role`,
      {
        roleName: `${props.config.lambdaFunctionName}-execution-role-for-snowflake`,
        assumedBy: new iam.CompositePrincipal(...snowflakePrincipals),
        externalIds: snowflakeExternalIds,
      }
    );

    const apiPolicyStatement = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ["execute-api:Invoke"],
      resources: ["execute-api:/*/POST/*"],
      principals: [
        new iam.ArnPrincipal(
          `arn:aws:sts::${props.config.awsAccountId}:assumed-role/${snowflakeRole.roleName}/snowflake`
        ),
      ],
    });
    const apiPolicy = new iam.PolicyDocument({
      statements: [apiPolicyStatement],
    });

    const restApi = new apigateway.RestApi(
      this,
      `${props.config.lambdaFunctionName}-rest-api`,
      {
        restApiName: `${props.config.lambdaFunctionName}-rest-api`,
        description: "Managed by CloudFormation (CDK)",
        endpointTypes: [apigateway.EndpointType.REGIONAL],
        deployOptions: { stageName: "default" },
        policy: apiPolicy,
      }
    );

    const integration = new apigateway.LambdaIntegration(lambdaFunction);
    const resouse = restApi.root;
    resouse.addMethod("POST", integration, {
      authorizationType: apigateway.AuthorizationType.IAM,
    });

    new cdk.CfnOutput(this, "SnowflakeApiAllowedPrefixes", {
      value: restApi.url,
      description: "api_allowed_prefixes",
    });
    new cdk.CfnOutput(this, "SnowflakeApiAwsRoleArn", {
      value: snowflakeRole.roleArn,
      description: "api_aws_role_arn",
    });
  }
}
