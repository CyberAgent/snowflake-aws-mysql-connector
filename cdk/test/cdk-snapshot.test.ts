import { SynthUtils } from "@aws-cdk/assert";
import * as cdk from "@aws-cdk/core";
import * as SnowflakeAWSMySQLConnector from "../lib/snowflake-aws-mysql-connector-stack";
import * as config from "../../src/conf/config";
import { SnowflakeAWSMySQLConnectorStackProps } from "../lib/snowflake-aws-mysql-connector-stack";

test("Snapshot Test", () => {
  const app = new cdk.App();
  const conf = config.getConfig("snapshot-test", "../../cdk/test/conf");
  const props: SnowflakeAWSMySQLConnectorStackProps = {
    config: conf,
    stackName: conf.stackName,
    lambdaAssetPath: "test/empty",
    lambdaLayerAssetPath: "test/empty",
    env: { region: conf.awsRegion, account: conf.awsAccountId },
  };
  // WHEN
  const stack = new SnowflakeAWSMySQLConnector.SnowflakeAWSMySQLConnectorStack(
    app,
    "SnowflakeAWSMySQLConnectorTestStack",
    props
  );
  // THEN
  const cloudFormation = SynthUtils.toCloudFormation(stack);
  expect(cloudFormation).toMatchSnapshot();
});
