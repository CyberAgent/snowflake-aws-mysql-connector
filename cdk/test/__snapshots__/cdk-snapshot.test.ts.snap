// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Test 1`] = `
Object {
  "Outputs": Object {
    "devsnowflakeawsmysqlconnectorrestapiEndpoint2C8E0A74": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "devsnowflakeawsmysqlconnectorrestapi034D486F",
            },
            ".execute-api.ap-northeast-1.",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "devsnowflakeawsmysqlconnectorrestapiDeploymentStagedefaultE3F69B5E",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24ArtifactHashBA91B77F": Object {
      "Description": "Artifact hash for asset \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3Bucket4D46ABB5": Object {
      "Description": "S3 bucket for asset \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861": Object {
      "Description": "S3 key for asset version \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParameters9237631b006453b8d2b275e61053dbf9c1ec0da492720b99a9d83ce3b802c656ArtifactHash03CBE45C": Object {
      "Description": "Artifact hash for asset \\"9237631b006453b8d2b275e61053dbf9c1ec0da492720b99a9d83ce3b802c656\\"",
      "Type": "String",
    },
    "AssetParameters9237631b006453b8d2b275e61053dbf9c1ec0da492720b99a9d83ce3b802c656S3BucketE83B8C93": Object {
      "Description": "S3 bucket for asset \\"9237631b006453b8d2b275e61053dbf9c1ec0da492720b99a9d83ce3b802c656\\"",
      "Type": "String",
    },
    "AssetParameters9237631b006453b8d2b275e61053dbf9c1ec0da492720b99a9d83ce3b802c656S3VersionKeyEE13941B": Object {
      "Description": "S3 key for asset version \\"9237631b006453b8d2b275e61053dbf9c1ec0da492720b99a9d83ce3b802c656\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3Bucket4D46ABB5",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "devsnowflakeawsmysqlconnectorA759905B": Object {
      "DependsOn": Array [
        "devsnowflakeawsmysqlconnectorServiceRole4253ECFA",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters9237631b006453b8d2b275e61053dbf9c1ec0da492720b99a9d83ce3b802c656S3BucketE83B8C93",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters9237631b006453b8d2b275e61053dbf9c1ec0da492720b99a9d83ce3b802c656S3VersionKeyEE13941B",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters9237631b006453b8d2b275e61053dbf9c1ec0da492720b99a9d83ce3b802c656S3VersionKeyEE13941B",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "STACK_ENV": "snapshot-test",
          },
        },
        "FunctionName": "dev-snowflake-aws-mysql-connector",
        "Handler": "index.handler",
        "Layers": Array [
          Object {
            "Ref": "devsnowflakeawsmysqlconnectorlayer9A47E73D",
          },
        ],
        "Role": Object {
          "Fn::GetAtt": Array [
            "devsnowflakeawsmysqlconnectorServiceRole4253ECFA",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 10,
        "VpcConfig": Object {
          "SecurityGroupIds": Array [
            "sg-abcdefghijk123",
          ],
          "SubnetIds": Array [
            "subnet-abcdefghijk123",
          ],
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "devsnowflakeawsmysqlconnectorLogRetentionE8A47EC3": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "devsnowflakeawsmysqlconnectorA759905B",
              },
            ],
          ],
        },
        "RetentionInDays": 3,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "devsnowflakeawsmysqlconnectorServiceRole4253ECFA": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "devsnowflakeawsmysqlconnectorlayer9A47E73D": Object {
      "Properties": Object {
        "CompatibleRuntimes": Array [
          "nodejs12.x",
        ],
        "Content": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters9237631b006453b8d2b275e61053dbf9c1ec0da492720b99a9d83ce3b802c656S3BucketE83B8C93",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters9237631b006453b8d2b275e61053dbf9c1ec0da492720b99a9d83ce3b802c656S3VersionKeyEE13941B",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters9237631b006453b8d2b275e61053dbf9c1ec0da492720b99a9d83ce3b802c656S3VersionKeyEE13941B",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "devsnowflakeawsmysqlconnectorrestapi034D486F": Object {
      "Properties": Object {
        "Description": "Managed by CloudFormation (CDK)",
        "EndpointConfiguration": Object {
          "Types": Array [
            "REGIONAL",
          ],
        },
        "Name": "dev-snowflake-aws-mysql-connector-rest-api",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "devsnowflakeawsmysqlconnectorrestapiAccount18867F40": Object {
      "DependsOn": Array [
        "devsnowflakeawsmysqlconnectorrestapi034D486F",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "devsnowflakeawsmysqlconnectorrestapiCloudWatchRole2E577E85",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "devsnowflakeawsmysqlconnectorrestapiCloudWatchRole2E577E85": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "devsnowflakeawsmysqlconnectorrestapiDeployment2D477E321c175f15134bf8c37d5ee732f898d875": Object {
      "DependsOn": Array [
        "devsnowflakeawsmysqlconnectorrestapipostPOSTF18BCB2E",
        "devsnowflakeawsmysqlconnectorrestapipostAA314BEF",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "devsnowflakeawsmysqlconnectorrestapi034D486F",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "devsnowflakeawsmysqlconnectorrestapiDeploymentStagedefaultE3F69B5E": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "devsnowflakeawsmysqlconnectorrestapiDeployment2D477E321c175f15134bf8c37d5ee732f898d875",
        },
        "RestApiId": Object {
          "Ref": "devsnowflakeawsmysqlconnectorrestapi034D486F",
        },
        "StageName": "default",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "devsnowflakeawsmysqlconnectorrestapipostAA314BEF": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "devsnowflakeawsmysqlconnectorrestapi034D486F",
            "RootResourceId",
          ],
        },
        "PathPart": "post",
        "RestApiId": Object {
          "Ref": "devsnowflakeawsmysqlconnectorrestapi034D486F",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "devsnowflakeawsmysqlconnectorrestapipostPOSTApiPermissionSnowflakeAWSMySQLConnectorTestStackdevsnowflakeawsmysqlconnectorrestapi0B44822CPOSTpostC3D7D6E9": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "devsnowflakeawsmysqlconnectorA759905B",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:ap-northeast-1:123456789012:",
              Object {
                "Ref": "devsnowflakeawsmysqlconnectorrestapi034D486F",
              },
              "/",
              Object {
                "Ref": "devsnowflakeawsmysqlconnectorrestapiDeploymentStagedefaultE3F69B5E",
              },
              "/POST/post",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "devsnowflakeawsmysqlconnectorrestapipostPOSTApiPermissionTestSnowflakeAWSMySQLConnectorTestStackdevsnowflakeawsmysqlconnectorrestapi0B44822CPOSTpost4F08307C": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "devsnowflakeawsmysqlconnectorA759905B",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:ap-northeast-1:123456789012:",
              Object {
                "Ref": "devsnowflakeawsmysqlconnectorrestapi034D486F",
              },
              "/test-invoke-stage/POST/post",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "devsnowflakeawsmysqlconnectorrestapipostPOSTF18BCB2E": Object {
      "Properties": Object {
        "AuthorizationType": "AWS_IAM",
        "HttpMethod": "POST",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:ap-northeast-1:lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "devsnowflakeawsmysqlconnectorA759905B",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "devsnowflakeawsmysqlconnectorrestapipostAA314BEF",
        },
        "RestApiId": Object {
          "Ref": "devsnowflakeawsmysqlconnectorrestapi034D486F",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
  },
}
`;
